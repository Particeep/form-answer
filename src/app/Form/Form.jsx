import "./Form.scss";

import React, {Component} from "react";
import {ExpensionStep, ExpensionStepper} from "../ExpensionStepper";
import Section from "./Section/Section";
import {connect} from "react-redux";
import formAction from "./formAction";

class Form extends Component {

    render() {
        return (
            <ExpensionStepper onNext={this.next} onEnd={this.end}>
                {this.props.form.sections.map(s =>
                    <ExpensionStep label="label!" component={<Section section={s}/>} key={s.id}/>
                )}
            </ExpensionStepper>
        );
    }

    componentWillMount() {
        this.props.dispatch(formAction.init({
            dateFormat: this.props.dateFormat,
            messages: this.props.messages,
            maxUploadFileSize: this.props.maxUploadFileSize,
            notifyChange: this.notifyChange,
            onUploadFile: this.onUploadFile,
        }));
    }

    onUploadFile = (sectionId, questionId, file) => {
        const {dispatch, onUploadFile} = this.props;
        onUploadFile(file, this.onFileUploaded(sectionId, questionId));
        dispatch(formAction.documentUploading(questionId, true));
    };

    onFileUploaded = (sectionId, questionId) => uploadedFile => {
        const {dispatch} = this.props;
        dispatch(formAction.documentUploading(questionId, false));
        dispatch(formAction.updateAnswer(questionId, [uploadedFile.name, uploadedFile.permalink]));
        dispatch(formAction.updateSectionValidity(sectionId, questionId, true));
    };

    notifyChange = (questionIdAnswered) => {
        setTimeout(() => this.props.onChange(
            {[questionIdAnswered]: this.props.answers[questionIdAnswered]},
            this.props.answers,
        ));
    };

    next = (sectionIndex) => {
        this.props.onSectionEnd(this.getSectionAnswers(sectionIndex));
    };

    end = () => {
        this.props.onSectionEnd(this.getSectionAnswers(this.props.form.sections.length - 1));
        this.props.onEnd(this.props.answers);
    };

    getSectionAnswers(sectionIndex) {
        const {answers} = this.props;
        const sectionQuestionIds = this.props.form.sections[sectionIndex].questions.map(q => q.id);
        return Object.keys(answers).filter(key => sectionQuestionIds.includes(key)).reduce((obj, key) => {
            obj[key] = answers[key];
            return obj;
        }, {});
    }
}

const state2Props = (state) => ({
    answers: state.form.answers,
});

export default connect(state2Props)(Form);