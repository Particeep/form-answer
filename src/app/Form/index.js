export {default as Form} from './Form';
export {default as formReducer} from './formReducer';