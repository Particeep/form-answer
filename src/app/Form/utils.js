import {questionType} from "./Question/Question";

export function mapSingleAnswer(answers) {
    return answers && answers[0];
}

export function parseSingleAnswer(answer) {
    if (answer || answer === '') return [answer];
    return undefined;
}

export function getDateFormatSeparator(dateFormat) {
    if (dateFormat.indexOf('/') >= 0) return '/';
    if (dateFormat.indexOf('-') >= 0) return '-';
    if (dateFormat.indexOf('.') >= 0) return '/';
    return '/';
}

export function isDependable(question) {
    return question.question_type === questionType.SELECT
        || question.question_type === questionType.RADIO;
}