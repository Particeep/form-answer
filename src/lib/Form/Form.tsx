import React from "react";
import {RecoilRoot} from "recoil";
import FormDisplay from "./FormDisplay";
import {FormProps} from "./FormProps";

const Form = (props: FormProps) => {
  return (
    <RecoilRoot>
      <FormDisplay {...props} />
    </RecoilRoot>
  )
}

export default Form
