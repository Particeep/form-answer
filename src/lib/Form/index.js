export {default as Form} from './Form';
export {default as Section} from './Section/Section';
export {formReducer} from './formReducer';