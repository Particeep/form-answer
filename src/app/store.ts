import {applyMiddleware, combineReducers, compose, createStore} from 'redux';
import thunk from 'redux-thunk';
import {formReducer} from '../lib/Form/index';
import {appReducer} from './app.reducer';

export const store = createStore(
  combineReducers({
    formAnswer: formReducer,
    app: appReducer,
  }),
  compose(
    applyMiddleware(thunk)
    // Redux DevToops Chrome plugins
    // , (<any> window).__REDUX_DEVTOOLS_EXTENSION__ && (<any> window).__REDUX_DEVTOOLS_EXTENSION__()
  )
);
