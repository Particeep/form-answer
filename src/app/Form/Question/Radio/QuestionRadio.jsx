import React, {Component} from "react";
import {FormControlLabel, Radio, RadioGroup} from "material-ui";
import {parseSingleAnswer} from "../../utils";
import {questionWrapper} from "../questionWrapper";
import {maxPossibilitiesBeforeAutocomplete} from "../Question";
import QuestionAutocomplete from "../Autocomplete/QuestionAutocomplete";

class QuestionRadio extends Component {

    render() {
        const {question, value} = this.props;
        if (question.possibilities.length < maxPossibilitiesBeforeAutocomplete)
            return (
                <RadioGroup
                    value={value}
                    onChange={e => this.handleChange(e.target.value)}
                >
                    {question.possibilities.map(p =>
                        <FormControlLabel
                            value={p.label}
                            control={<Radio/>}
                            label={p.label}
                            key={p.id}
                            onClick={() => value === p.label && this.handleChange('')}/>
                    )}
                </RadioGroup>
            );
        return <QuestionAutocomplete {...this.props}/>
    }

    handleChange = value => {
        this.props.onChange(value);
        const possibility = this.props.question.possibilities.find(p => p.label === value);
        if (possibility) this.props.onCheckPossibility(possibility.id);
    };
}

export default questionWrapper(QuestionRadio);