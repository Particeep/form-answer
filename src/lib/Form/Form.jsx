import "./Form.scss";

import React, {Component} from "react";
import {ExpensionStep, ExpensionStepper} from "../ExpensionStepper";
import {Section} from "./Section";
import {connect} from "react-redux";
import {formAction} from "./formAction";
import {isFunction} from "../utils/common";
import {ApiParser} from "../utils/ApiParser";
import {questionType} from "./Question/QuestionType";

class Form extends Component {

    render() {
        return (
            <ExpensionStepper free={this.props.readonly} onNext={this.next} onEnd={this.end}>
                {this.props.form.sections.map(s =>
                    <ExpensionStep label={s.name} component={<Section section={s}/>} key={s.id}/>
                )}
            </ExpensionStepper>
        );
    }

    componentWillMount() {
        this.parser = new ApiParser(this.props.dateFormat);
        this.initReducer();
        this.initAnswers();
    }

    initReducer() {
        const {
            dispatch,
            dateFormat,
            messages,
            maxUploadFileSize,
            readonly,
        } = this.props;
        dispatch(formAction.init({
            dateFormat: dateFormat,
            messages: messages,
            maxUploadFileSize: maxUploadFileSize,
            notifyChange: this.onChange,
            onUploadFile: this.onUploadFile,
            readonly: readonly || false,
        }));
    }

    initAnswers() {
        this.props.form.sections.forEach(s => s.questions.forEach(q => {
            if (q.question_type === questionType.LABEL) return;
            this.props.dispatch(formAction.updateAnswer(
                q.id,
                q.question_type,
                this.parser.fromApi(q.question_type)(q.answers)
            ))
        }));
    }

    onUploadFile = (sectionId, questionId, file) => {
        const {dispatch, onUploadFile} = this.props;
        onUploadFile(file, this.onFileUploaded(sectionId, questionId));
        dispatch(formAction.documentUploading(questionId, true));
    };

    onFileUploaded = (sectionId, questionId) => uploadedFile => {
        const {dispatch} = this.props;
        dispatch(formAction.documentUploading(questionId, false));
        dispatch(formAction.updateAnswer(questionId, questionType.DOCUMENT, [uploadedFile.name, uploadedFile.permalink]));
        dispatch(formAction.updateSectionValidity(sectionId, questionId, true));
        this.onChange(questionId);
    };

    onChange = (questionIdAnswered) => {
        if (!isFunction(this.props.onChange)) return;
        setTimeout(() =>
            this.props.onChange(this.parseAnswer(questionIdAnswered, this.props.answers[questionIdAnswered]))
        );
    };

    next = (sectionIndex) => {
        if (!isFunction(this.props.onSectionEnd)) return;
        this.props.onSectionEnd(this.parseAnswers(this.getSectionAnswers(sectionIndex)));
    };

    end = () => {
        const {form, answers, onSectionEnd, onEnd} = this.props;
        if (isFunction(onSectionEnd))
            onSectionEnd(this.parseAnswers(this.getSectionAnswers(form.sections.length - 1)));
        if (isFunction(onEnd))
            onEnd(this.parseAnswers(answers));
    };

    getSectionAnswers(sectionIndex) {
        const {answers} = this.props;
        const sectionQuestionIds = this.props.form.sections[sectionIndex].questions.map(q => q.id);
        return Object.keys(answers).filter(key => sectionQuestionIds.includes(key)).reduce((obj, key) => {
            obj[key] = answers[key];
            return obj;
        }, {});
    }

    parseAnswers = (answers) => {
        return Object.keys(answers).map(k => this.parseAnswer(k, answers[k])).filter(v => v);
    };

    parseAnswer = (id, answer) => {
        const value = this.parser.toApi(answer.type)(answer.value);
        if (value)
            return {question_id: id, answer: value}
    };
}

const state2Props = (state) => ({
    answers: state.formAnswer.answers,
});

export default connect(state2Props)(Form);