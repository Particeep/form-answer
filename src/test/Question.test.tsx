import * as React from 'react';
import {Provider} from 'react-redux';
import {form, s1_qRADIO} from './utils/form';
import {Form} from '../lib/Form';
import {testStore} from './utils/store';
import {shallow} from 'enzyme';
import QuestionRadio from '../lib/Form/Question/Radio/QuestionRadio';
import {shallow} from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import * as enzyme from 'enzyme';

test('Question TEXT', () => {
    const $form = shallow(
        <Provider store={testStore}>
            <Form form={form(s1_qRADIO)}/>
        </Provider>
    );
    console.log($form.find(QuestionRadio).length);
});