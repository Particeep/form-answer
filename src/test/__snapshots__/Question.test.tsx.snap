// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question TEXT 1`] = `
<main>
  <main
    className="ExpensionStep -current"
  >
    <header
      className="ExpensionStep_header"
      onClick={[Function]}
    >
      1
      . 
      Sélection de l’appel à candidature et des leviers d’optimisation 
    </header>
    <div
      className="MuiCollapse-container-1 MuiCollapse-entered-2 ExpensionStep_body"
      style={
        Object {
          "minHeight": "0px",
        }
      }
    >
      <div
        className="MuiCollapse-wrapper-3"
      >
        <div
          className="MuiCollapse-wrapperInner-4"
        >
          <div
            className="ExpensionStep_content"
          >
            <main>
              <div
                className="Section_label"
              />
              <main
                className="Question Question-Q1"
              >
                <div
                  className="Question_label"
                >
                  Appel à candidature obligatoire :
                  <span
                    className="Question_required"
                  >
                    *
                  </span>
                </div>
                <div
                  className="Question_description"
                />
                <div
                  className="Question_answer"
                >
                  <div
                    className="MuiFormGroup-root-5"
                    role="radiogroup"
                  >
                    <label
                      className="MuiFormControlLabel-root-7"
                      onClick={[Function]}
                    >
                      <span
                        className="MuiButtonBase-root-26 MuiIconButton-root-20 MuiSwitchBase-root-15 MuiSwitchBase-default-17 MuiRadio-default-10 MuiSwitchBase-checked-18 MuiRadio-checked-11 MuiRadio-checkedSecondary-13"
                        onBlur={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                        onMouseDown={[Function]}
                        onMouseLeave={[Function]}
                        onMouseUp={[Function]}
                        onTouchEnd={[Function]}
                        onTouchMove={[Function]}
                        onTouchStart={[Function]}
                        tabIndex={null}
                      >
                        <span
                          className="MuiIconButton-label-25"
                        >
                          <svg
                            aria-hidden="true"
                            className="MuiSvgIcon-root-29"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M12 7c-2.76 0-5 2.24-5 5s2.24 5 5 5 5-2.24 5-5-2.24-5-5-5zm0-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                            />
                          </svg>
                          <input
                            checked={true}
                            className="MuiSwitchBase-input-16"
                            disabled={false}
                            onChange={[Function]}
                            type="radio"
                            value="Extension des consignes de tri"
                          />
                        </span>
                        <span
                          className="MuiTouchRipple-root-35"
                        />
                      </span>
                      <span
                        className="MuiTypography-root-42 MuiTypography-body1-51 MuiFormControlLabel-label-9"
                      >
                        Extension des consignes de tri
                      </span>
                    </label>
                    <label
                      className="MuiFormControlLabel-root-7"
                      onClick={[Function]}
                    >
                      <span
                        className="MuiButtonBase-root-26 MuiIconButton-root-20 MuiSwitchBase-root-15 MuiSwitchBase-default-17 MuiRadio-default-10"
                        onBlur={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                        onMouseDown={[Function]}
                        onMouseLeave={[Function]}
                        onMouseUp={[Function]}
                        onTouchEnd={[Function]}
                        onTouchMove={[Function]}
                        onTouchStart={[Function]}
                        tabIndex={null}
                      >
                        <span
                          className="MuiIconButton-label-25"
                        >
                          <svg
                            aria-hidden="true"
                            className="MuiSvgIcon-root-29"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                            />
                          </svg>
                          <input
                            checked={false}
                            className="MuiSwitchBase-input-16"
                            disabled={false}
                            onChange={[Function]}
                            type="radio"
                            value="Ne pas laisser couler l'eau"
                          />
                        </span>
                        <span
                          className="MuiTouchRipple-root-35"
                        />
                      </span>
                      <span
                        className="MuiTypography-root-42 MuiTypography-body1-51 MuiFormControlLabel-label-9"
                      >
                        Ne pas laisser couler l'eau
                      </span>
                    </label>
                  </div>
                </div>
              </main>
              <div
                className="Section_action"
              >
                <button
                  className="MuiButtonBase-root-26 MuiButton-root-66 MuiButton-raised-71 MuiButton-raisedPrimary-73 Section_end"
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex="0"
                  type="button"
                >
                  <span
                    className="MuiButton-label-67"
                  >
                    End
                  </span>
                  <span
                    className="MuiTouchRipple-root-35"
                  />
                </button>
              </div>
            </main>
          </div>
        </div>
      </div>
    </div>
  </main>
</main>
`;
